DEVOPS-591 warn users on deletion

redondos opened the pull request:
Sample usage:
```
booster 10gen-ops/ksec ‹redondos/DEVOPS-591*› » ksec delete angelo-test                                                                                                                                                                                                             (aws:platform)(k8s:test.k8s.local:devops)
Are you sure? This action cannot be reversed. (y/n) [n]:
booster 10gen-ops/ksec ‹redondos/DEVOPS-591*› » ksec delete angelo-test                                                                                                                                                                                                             (aws:platform)(k8s:test.k8s.local:devops)
Are you sure? This action cannot be reversed. (y/n) [n]: y
Deleted secret "angelo-test"
```

Test run (no changes):
```
booster 10gen-ops/ksec ‹redondos/DEVOPS-591*› » make test                                                                                                                                                                                                                           (aws:platform)(k8s:test.k8s.local:devops)
go vet ./...
go test -cover ./...
ok  	github.com/10gen-ops/ksec/cmd/ksec	(cached)	coverage: 54.7% of statements
ok  	github.com/10gen-ops/ksec/pkg/models	(cached)	coverage: 55.3% of statements
?   	github.com/10gen-ops/ksec/pkg/version	[no test files]
```

richerve commented:
LGTM, but I'll let Colin approve or not.

redondos commented:
FYI: Since this is a fork of `colinhoglund/ksec` Github assumed I was trying to submit a PR to the parent repo. Bear this in mind when creating PRs against this codebase since the link below will redirect and you might mistakenly create it in the wrong org.

```
Enumerating objects: 21, done.
Counting objects: 100% (21/21), done.
Delta compression using up to 8 threads
Compressing objects: 100% (12/12), done.
Writing objects: 100% (14/14), 6.11 KiB | 6.11 MiB/s, done.
Total 14 (delta 8), reused 0 (delta 0)
remote: Resolving deltas: 100% (8/8), completed with 5 local objects.
remote:
remote: Create a pull request for 'redondos/DEVOPS-591' on GitHub by visiting:
remote:      https://github.com/10gen-ops/ksec/pull/new/redondos/DEVOPS-591
remote:
To github.com:10gen-ops/ksec
 * [new branch]      redondos/DEVOPS-591 -> redondos/DEVOPS-591
Branch 'redondos/DEVOPS-591' set up to track remote branch 'redondos/DEVOPS-591' from 'origin'.
```

colinhoglund commented on cmd/ksec/delete.go:
```diff
@@ -3,6 +3,7 @@ package main
 import (
 	"fmt"
 
+	"github.com/Songmu/prompter"
```
This pulls in a handful of new dependencies for what is pretty simple to do with a `bufio.NewScanner(os.Stdin)` (which is indeed what is happening in prompter). I'd rather not have to review these dependencies if we don't have to. Can we pull this out?

dskatz commented on cmd/ksec/delete.go:
```diff
@@ -14,6 +15,16 @@ var deleteCmd = &cobra.Command{
 }
 
 func deleteCommand(cmd *cobra.Command, args []string) error {
+	noconfirm, err := cmd.Flags().GetBool("yes")
```
Agreed. Or just a var name of yes, to match the flag.

colinhoglund commented on cmd/ksec/delete.go:
```diff
@@ -14,6 +14,16 @@ var deleteCmd = &cobra.Command{
+		return err
+	}
+
+	if !skipconfirm && !askConfirmation("Are you sure? This action cannot be reversed.") {
```
This confirmation shows up even if the secret doesn't exist. Probably fine, but slightly confusing.

redondos replied:
Sound good, I made it cancel if any of the specified secrets don't exist. Could do the opposite, but this seemed appropriate.

colinhoglund replied:
This is better. I still think the workflow could provide more feedback for various edge cases. More tests would help uncover these. Does it make more sense to just wrap the whole `Get, Confirm, Delete` workflow in a single loop?

This kind of mimics the workflow of similar tools like `rm` or `kubectl delete secret`:

```
 ~ $ rm fake bogus
rm: fake: No such file or directory
rm: bogus: No such file or directory
 ~ $ kubectl delete secret fake bogus
Error from server (NotFound): secrets "fake" not found
Error from server (NotFound): secrets "bogus" not found
 ~ $ touch fake bogus
 ~ $ grm -i fake bogus
grm: remove regular empty file 'fake'? y
grm: remove regular empty file 'bogus'? y
 ~ $ ksec create bogus
Created secret "bogus"
 ~ $ kubectl delete secret fake bogus
secret "bogus" deleted
Error from server (NotFound): secrets "fake" not found
 ~ $ ksec create fake
Created secret "fake"
 ~ $ kubectl delete secret fake bogus
secret "fake" deleted
Error from server (NotFound): secrets "bogus" not found
```

dskatz replied:
+1

colinhoglund commented on cmd/ksec/utils.go:
```diff
@@ -13,3 +14,20 @@ func outputTabular(lines []string) {
+	scanner.Scan()
+	response := scanner.Text()
+	if scanner.Err() != nil {
+		return false
```
This conditional seems redundant since default behavior is to return false and this isn't surfacing the error to the user. Should probably either surface the error or just not bother.

redondos replied:
True, I'm showing the error and exiting now. Do you prefer it to return instead?

redondos commented:
> Existing tests are failing, and we should also create some new testing to validate this logic.

Fixed the existing test. What extra logic are you thinking?

colinhoglund commented on cmd/ksec/delete.go:
```diff
@@ -14,6 +14,23 @@ var deleteCmd = &cobra.Command{
+
+	for _, name := range args {
+		if _, err := secretsClient.Get(name); err != nil {
+			fmt.Println(err)
```
you can just `return err` and this will get handled upstream as the other commands are doing.

dskatz replied:
+1. Its more consistent with the rest of the function.

colinhoglund replied:
Or if we wrap `Get, Confirm, Delete` in a single loop, you could also just print the err to stderr and continue.

colinhoglund commented on cmd/ksec/utils.go:
```diff
@@ -13,3 +14,21 @@ func outputTabular(lines []string) {
+	scanner.Scan()
+	response := scanner.Text()
+	if err := scanner.Err(); err != nil {
+		fmt.Fprintln(os.Stderr, "invalid input: ", err)
```
Not sure if `invalid input: ` is necessary. Wouldn't something similar be in the `err` message? Should this just be `fmt.Fprintf(os.Stderr, err.Error())`?

colinhoglund commented:
Addressed comments.

Example workflow
```
 ~ $ ksec create colin colin1
Created secret "colin"
Created secret "colin1"
 ~ $ ksec delete colin colin1
Delete secret "colin"? This action cannot be reversed. [y/N]: n
Delete canceled
Delete secret "colin1"? This action cannot be reversed. [y/N]: n
Delete canceled
 ~ $ ksec delete colin colin1
Delete secret "colin"? This action cannot be reversed. [y/N]: y
Deleted secret "colin"
Delete secret "colin1"? This action cannot be reversed. [y/N]: n
Delete canceled
 ~ $ ksec delete colin colin1
secrets "colin" not found
Delete secret "colin1"? This action cannot be reversed. [y/N]: y
Deleted secret "colin1"
 ~ $ ksec delete colin colin1
secrets "colin" not found
secrets "colin1" not found
```

redondos commented:
> Addressed comments.
> 
> Example workflow
> 
> ```
>  ~ $ ksec create colin colin1
> Created secret "colin"
> Created secret "colin1"
>  ~ $ ksec delete colin colin1
> Delete secret "colin"? This action cannot be reversed. [y/N]: n
> Delete canceled
> Delete secret "colin1"? This action cannot be reversed. [y/N]: n
> Delete canceled
>  ~ $ ksec delete colin colin1
> Delete secret "colin"? This action cannot be reversed. [y/N]: y
> Deleted secret "colin"
> Delete secret "colin1"? This action cannot be reversed. [y/N]: n
> Delete canceled
>  ~ $ ksec delete colin colin1
> secrets "colin" not found
> Delete secret "colin1"? This action cannot be reversed. [y/N]: y
> Deleted secret "colin1"
>  ~ $ ksec delete colin colin1
> secrets "colin" not found
> secrets "colin1" not found
> ```

Thanks for taking this over. Defaulting to `-i` seems fine, and safer than asking for confirmation just once for all items.

However a `-f` flag would be nice in the future, in case users want to invoke this from a script without having to resort to any input tricks.

richerve commented:
Completely agree with having an `-f` in this case.

colinhoglund commented:
> a `-f` flag would be nice

Not sure I follow...

The whole purpose of this PR is to default to ask for confirmation and add a `-y` flag that forces deletion. Unclear what a `-f` flag would do.

```
$ ksec delete -y colin colin1
Deleted secret "colin"
Deleted secret "colin1"
```

richerve commented:
> > a `-f` flag would be nice
> 
> Not sure I follow...
> 
> The whole purpose of this PR is to default to ask for confirmation and add a `-y` flag that forces deletion. Unclear what a `-f` flag would do.
> 
> ```
> $ ksec delete -y colin colin1
> Deleted secret "colin"
> Deleted secret "colin1"
> ```

Sorry, for me `-f` is the same as `-y` in your example

redondos commented:
> > a `-f` flag would be nice
> 
> Not sure I follow...
> 
> The whole purpose of this PR is to default to ask for confirmation and add a `-y` flag that forces deletion. Unclear what a `-f` flag would do.
> 
> ```
> $ ksec delete -y colin colin1
> Deleted secret "colin"
> Deleted secret "colin1"
> ```

🤦‍♂  I forgot I added a `-y`. Let's merge this once and for all :)
